sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -508,164 +498,202 @@ export class EventHubSender extends LinkEntity {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender "${this.name}", ` +
+                `received a modified disposition.Hence we are rejecting the promise.`
             );
-            const err = translate(context!.delivery!.remote_state!.error);
-            log.error(err);
-            reject(err);
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const removeListeners = (): void => {
+          // When `removeListeners` is called on timeout, the sender might be closed and cleared
+          // So, check if it exists, before removing listeners from it.
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAborted);
+          }
+          if (this._sender) {
+            this._sender.removeListener(SenderEvents.rejected, onRejected);
+            this._sender.removeListener(SenderEvents.accepted, onAccepted);
+            this._sender.removeListener(SenderEvents.released, onReleased);
+            this._sender.removeListener(SenderEvents.modified, onModified);
+          }
+        };
+
+        const rejectOnSendError = (err: Error) => {
+          err = translate(err);
+          log.error(
+            "[%s] An error occurred while performing send on %s",
+            this._context.connectionId,
+            this.name,
+            err
+          );
+          reject(err);
+        };
+
+        const actionAfterTimeout = () => {
+          clearTimeout(waitTimer);
+          removeListeners();
+          const desc: string =
+            `[${this._context.connectionId}] Sender "${this.name}" with ` +
+            `address "${this.address}", was not able to send the message right now, due ` +
+            `to operation timeout.`;
+          log.error(desc);
+          const e: Error = {
+            name: "OperationTimeoutError",
+            message: desc
           };
-          onReleased = (context: EventContext) => {
-            removeListeners();
-            log.error(
-              "[%s] Sender '%s', got event released.",
-              this._context.connectionId,
-              this.name
-            );
-            let err: Error;
-            if (context!.delivery!.remote_state!.error) {
-              err = translate(context!.delivery!.remote_state!.error);
-            } else {
-              err = new Error(
-                `[${this._context.connectionId}] Sender '${this.name}', ` +
-                  `received a release disposition.Hence we are rejecting the promise.`
-              );
+          return reject(translate(e));
+        };
+
+        const sendEventPromise = () =>
```

Comment:
I only see this function being called if the sender link wasn't already initialized. If the link is open, it doesn't look like this gets called at all.

Reply:
That would mean that the second send() would not send anything.

Also, the `resolve` callback for this promise is not called from any code path. So, how does this promise get resolved?

Reply:
Oops, yes - fixed checks. Actually, this got simplified as well with the async-await suggestion.

Reply:
resolve was being called by the `onAccepted` feedback. That was just moved out so TypeScript must have marked it as not in use

Reply:
The resolve called by `onAccepted` is the resolve on the outer promise. I am referring to the inner promise that is not getting resolved

Reply:
The resolve called by `onAccepted` is the resolve on the outer promise. I am referring to the inner promise that is not getting resolved. Even if scoping takes care of using the right resolve, for the ease of code maintainability it would be better if
- The inner promise deals with the code for the  4 callbacks of sender
- The outer promise deals with the timer, and the code for the callback of the timer (including `removeListeners`)